rfcs/20190815-tfdbg-v2-callbacks.md

```diff
@@ -0,0 +1,356 @@
+  #
+  #   The third (last) invocation of the callback is due to the eager execution
+  #   of the FuncGraph, with the `op_type` arg being `tf.Graph`, the `op_name`
+  #   arg being something like `_inference_log_1plusp_30`, and the `outputs` arg
+  #   being an EagerTensor of value 1.386 (≈log(4.0)).
+```
+
+The example above dealt with a relatively simple FuncGraph. But the proposed API
+applies to more complex FuncGraphs, including those that involve control flow
+such as if-else and while loops. For instance, see the code example below.
+
+```python
+with tf.debugging.op_callback(my_callback):
+
+  @dtf.function
+  def collatz(x):
+    n = tf.convert_to_tensor((0,))
+    while x != 1:
+      n += 1
+      if x % 2 == 0:
+        x = x // 2
+      else:
+        x = 3 * x + 1
+    return n
+
+  y = collatz(tf.constant(42))
+  # ↑ During the execution of the line above, `my_callback` is invoked
+  #   for the creation of all ops in the TF While loop's body and condition,
+  #   and the TF Cond op's branches. In addition, `my_callback` will be
+  #   invoked for the runtime execution of the FuncGraph converted from
+  #   `collatz`.
+```
+
+### Runtime Instrumentation of Ops in FuncGraphs
+
+This API also supports the runtime instrumentation of ops in FuncGraphs. To see
+why, realize the fact that the return values of the callback will override the
+original Tensors for downstream graph construction. Hence, new ops can be
+created inside the body of the callback. (The ops created inside the calbacks
+themselves will be skipped the callback, which avoids infinite loops.) Such new
+ops can consume the original output Tensors of the op and generate new output
+Tensors for the callback function to return. This workflow results in debugging
+nodes being inserted in the op’s graph.
+
+The two code snippets listed below (labeled "Listing 1" and "Listing 2") provide
+a concrete example of the workflow involved. The code in Listing 1 uses
+@tf.function to construct a simple TF graph. In particular, the Python code in
+my_func() is converted into a FuncGraph. The code in Listing 2 performs the same
+action, but with the code placed inside an `tf.debugging.op_callback()` scope.
+The callback function passed to the context manager constructs a debug op
+(DebugIdentityV2) for each of the op’s output Tensor and returns the output
+tensors of these debug ops, the result of which is shown in Panel B of the
+figure below: each tensor now has a DebugIdentityV2 op attached to it.
+
+```python
+# Listing 1. Define a FuncGraph without `tf.debugging.op_callback()`.
+
+@tf.function
+def my_func(x):
+  return tf.math.log(1 + x)
+```
+
+```python
+# Listing 2. Define a FuncGraph with `tf.debugging.op_callback()` and
+#            by passing a callback that overrides the symbolic output
+#            tensors of the just-created graph ops.
+
+def debugger_callback(op_type, inputs, attrs, outputs,
+                      op_name=None, graph=None):
```

Comment:
Is there a set of built-in callbacks that would be useful to provide?

Reply:
Yes, e.g., 
1. a callback that crashes with a helpful error message and stack trace when any output tensor contains NaN or Infinity, including during eager execution and execution of intra-graph ops.
2. a callback that dumps numeric summary of all tenors (or a subset of them) to the filesystem during model training, to facilitate post hoc analysis of failed training jobs
3. a callback that sends debug info to a debug server via a grpc stream, to facilitate interactive debugging

But the detailed design is outside the scope of this doc. This is briefly mentioned in the newly added "What this callback API offers beyond pdb" section.